keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+
+The main characteristic of this controller is the fact that this controller is deployed on all the nodes in the cluster using a `DaemonSet`, and that each instance is responsible for managing `Task` targeting a specific node.
+
+`Task` reconciliation triggers the execution of kubeadm commands on such node; in order to do so, `Task controller` require to be executed in privileged mode.
+
+#### Observability and Operability
+The kubeadm operator is designed to take responsibility for actions that as of today are in charge of users, and we want to ensure that the users can feel comfortable in handling over such responsibility.
+
+This goal influences the kubeadm operator design as described in the following paragraphs
+
+##### Execution order
+
+The kubeadm operator is going to execute `Action`'s `Task`s in a predictable and consistent order:
+
+- `TaskGroup` will be generated in the order encoded in each action and always executed in sequential order.
+- If a `TaskGroup` uses `sequential` policy, `Task` will be executed according to the alphabetical order of Nodes.
+- Additional `TaskGroup` policies introduced in future, are expected to introduce similar guarantees. 
+
+Additionally, it will be possible to execute actions in `dry-run` mode, thus allowing users to have a preview of the resulting task order.
+
+##### Execution modes
+
+The kubeadm operator is going to execute`Task`s acconding to the order described above, and by default a new `Task` will be automatically started as soon as the previous task completes.
+
+At any time during the execution, the user will be allowed to pause the `Action` execution using the new `kubeadm operator pause` command.
+Similarly, it will be possible to restart the execution with `kubeadm operator restart`.
+
+It will be also possible to execute action in `Controlled` mode; with this option, the kubeadm operator to automatically pause before actually executing each task; this will allow the user to check the task details and decide whether to proceed or not.
+
+> NB. Once a `Task` is handled over to kubeadm, it is not possible to pause its execution; that means that `kubeadm operator pause` does not affect running `Task`, but it prevents new new `Task` to be started. 
+
+##### Error management
+
+In case of errors, the kubeadm operator will immediately pause `Action` execution.
+
+During the proof of concept we will explore options for allowing retry of failed tasks and/or restart and skipping failed task.
+
+
+#### Extensibility
+
+The kubeadm operator provides high level `Action` encoding the knowledge of how kubeadm workflows should be performed in the form of `TaskGroup` and `Task`. 
+
+However, such out of the box experience might not be fit for advanced use cases, custom clusters etc.
+
+The kubeadm operator will provide a toolbox for allowing users to address such use cases as well:
+
+- It will be introduced a `CustomAction` type, allowing the user to define custom lists of `TaskGroup` and `Task`.
```

Comment:
I think this somewhat fails out of the scope of this RFC? Once that the kubeadm operator is in place we can check for extensibility and how to customize it in the best way possible, once that we have feedback from users. I would try to aim for a MVP regarding the operator, then iterate.

Reply:
Added a sentence to formalize this is a stretched objective for the MVP.
However, I'm confident we can get this almost for free and it helps for testing the operator itself, so currently I'm still for getting this at the very beginning